src/main/java/com/emc/pravega/perf/ReaderWorker.java

```diff
@@ -0,0 +1,83 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.emc.pravega.perf;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * abstract class for Readers.
+ */
+public abstract class ReaderWorker extends Worker implements Callable<Void> {
+
+    ReaderWorker(int readerId, int eventsPerWorker, int secondsToRun, Instant start,
+                 PerfStats stats, String readergrp, int timeout) {
+        super(readerId, eventsPerWorker, secondsToRun,
+                false, 0, start,
+                stats, readergrp, timeout);
+    }
+
+    /**
+     * read the data.
+     */
+    public abstract String readData();
+
+    /**
+     * close the consumer/reader.
+     */
+    public abstract void close();
+
+    /**
+     * read all the data from reader/consumer.
+     *
+     * @param drainStats object which collects the details about the data read till there is no data to read from stream
+     */
+    public void cleanupEvents(PerfStats drainStats) {
+        String ret = null;
+        do {
+            final Instant startTime = Instant.now();
+            ret = readData();
+            if (ret != null) {
+                drainStats.recordTime(null, startTime, ret.length());
+            }
+        } while (ret != null);
+        drainStats.printTotal(Instant.now());
+    }
+
+    @Override
+    public Void call() {
+        String ret = null;
+        try {
+
+            for (int i = 0; (i < eventsPerWorker) &&
```

Comment:
Should readers be capped based on a count per worker? The events won't be uniformly distributed, so if some exit early because they hit the count it's going to mean some events go unread. Perhaps readers should just be a time thing, or perhaps they could have some sort of 'stop()' method. (If you want to do that in a different PR that's fine)

Reply:
yes, currently I am just making sure that reader and transaction works fine but i am not publishing the performance numbers of reader and transactions.  once we close the pravega writer performance v/s kafka writer , I will revisit this code again.